common/protob/messages-cardano.proto

```diff
@@ -218,6 +241,24 @@ message CardanoSignTx {
         repeated uint32 path = 1;
         required uint64 amount = 2;
     }
+
+    message CardanoCatalystRegistrationParametersType {
+        required bytes voting_public_key = 1;
+        repeated uint32 staking_path = 2;
+        required CardanoAddressParametersType reward_address_parameters = 3;
+        required uint64 nonce = 4;
+    }
+
+    message CardanoTxMetadataType {
+        required CardanoMetadataType type = 1;
+        optional CardanoCatalystRegistrationParametersType catalyst_registration_parameters = 2;
+    }
+
+    message CardanoTxAuxiliaryDataType {
```

Comment:
reading through the details, it seems that you expect more auxiliary data types, of which "metadata" is the only one supported right now?

Reply:
According to the [CDDL spec](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley-ma/shelley-ma-test/cddl-files/shelley-ma.cddl#L215) `auxiliary_data` can even now contain `auxiliary_scripts` - which we don't need to support here yet. `auxiliary_scripts` is thus one field that might be added to `CardanoTxAuxiliaryDataType` later, but more might be potentially introduced. The `metadata` and `auxiliary_scripts` fields can be both present at the same type, so they aren't exclusive types of `auxiliary_data`. So when it comes to `CardanoTxAuxiliaryDataType` only new fields are probably going to be added, not `CardanoAuxiliaryDataType`s.

However, when it comes to `CardanoTxMetadataType`, we expect this to expand further with other `CardanoMetadataType`s and other optional fields.

Thus it seemed best to us to separate it this way so both can easily be extended without bloating `CardanoSignTx`.

> I'm wondering if we could also leave out blob here and keep the metadata field in the original transaction?

This could be a bit misleading, since what gets sent in the `blob` is now `auxiliary_data` not just the `metadata` part of it. So if someone would send `metadata` and `auxiliary_data` at the same time, expecting `auxiliary_data` to be built from these two parts, only one of those would actually be added to the transaction. Of course, we could validate it, but it seems confusing. This way the client's intent is explicit even at the price of backwards compatibility. Also, metadata probably haven't been widely used yet so it shouldn't be too much trouble.

Reply:
On second thought - perhaps it could indeed be simplified. We're going to have an internal discussion on the topic and I'll get back to you.

Reply:
> So if someone would send `metadata` and `auxiliary_data` at the same time, expecting `auxiliary_data` to be built from these two parts, only one of those would actually be added to the transaction.

the way I'm reading it, you're **either** sending a blob that represents all of the auxiliary data, **or** some structured data. Which actually seems to be an argument _in favor_ of keeping the old field. True, you need to check that the host does not send both, but that's what I'd expect.
We can (and probably should) rename the `metadata` field to `auxiliary_data_blob`, which breaks API compatibility (can't use old app with new Connect) but not wire compatibility (old app with old Connect still works)

Reply:
We discussed this and ended up with the simplification you proposed. Dropping `CardanoTxMetadataType` and moving `catalyst_registration_parameters` to `CardanoTxAuxiliaryDataType` and also removing `CardanoAuxiliaryDataType` and serializing based on whether `blob` or `catalyst_registration_parameters` is filled.

However, we would still remove the `metadata` field from the transaction and move it to `CardanoTxAuxiliaryDataType`. This way there's only one field related to auxiliary data in the transaction and it makes it more obvious that the fields are somewhat related. The impact on existing users should be close to zero as metadata haven't been widely used by any major Cardano tools.

Reply:
:+1: go ahead

Reply:
Restructured in https://github.com/trezor/trezor-firmware/pull/1564/commits/c82d4d141bda52e6da40a9fcb938a1830d43e5e8.